// tsconfig-to-dual-package/module/cli.js
import { parseArgs } from "node:util";
import { tsconfigToDualPackages } from "./tsconfig-to-dual-package.js";
const HELP = `
    Usage
      $ tsconfig-to-dual-package [Option] <tsconfig.json>
 
    Options
      --cwd                 [String] current working directory. Default: process.cwd()
      --debug               [Boolean] Enable debug output
      --help                [Boolean] show help

    Examples
      # Find tsconfig*.json in cwd and convert to dual package
      $ tsconfig-to-dual-package
      # Convert specified tsconfig.json to dual package
      $ tsconfig-to-dual-package ./config/tsconfig.esm.json ./config/tsconfig.cjs.json

`;
export const createCli = () => {
    return parseArgs({
        strict: true,
        allowPositionals: true,
        options: {
            cwd: {
                type: "string",
                default: process.cwd()
            },
            help: {
                type: "boolean",
                alias: "h"
            },
            debug: {
                type: "boolean",
                default: false
            }
        }
    });
};
export const run = async (cli = createCli()) => {
    var _a;
    if (cli.values.help) {
        return { exitStatus: 0, stdout: HELP, stderr: null };
    }
    const options = {
        targetTsConfigFilePaths: cli.positionals,
        cwd: (_a = cli.values.cwd) !== null && _a !== void 0 ? _a : process.cwd(),
        debug: cli.values.debug
    };
    await tsconfigToDualPackages(options);
    return {
        stdout: null,
        stderr: null,
        exitStatus: 0
    };
};
